### PR TITLE
GSW-1407-unify-panic-msg-with-error-code

### PR DESCRIPTION
**Pull request by r3v4s**
## refactor
1. unify panic msg with pre-defined error code

### CODE REVIEW

**Review by onlyhyde — DISMISSED** [review]

**Review by notJoon — DISMISSED** [review]

**Review by dongwon8247 — CHANGES_REQUESTED** [review]
I will review in depth soon.

But before that please make sure to connect all error messages within the logic and include all test cases.

**Comment by dongwon8247** [comment]
The goal of this is not only to clean the codebase but also to show error messages in the GnoSwap interface.
Has this been reviewed with @jinoosss? if not please do.

File: community_pool/errors.gno
Comment (dongwon8247) [inline community_pool/errors.gno:12]: Will these need to stay here if we remove the current grc20registry and replace it with /r/grc20reg?

File: gov/governance/errors.gno
Comment (dongwon8247) [inline gov/governance/errors.gno:27]:
```suggestion
	errProposalQuorumNotSatisfied                     = errors.New("[GNOSWAP-GOVERNANCE-018] proposal quorum not met")
```

File: gov/governance/errors.gno
Comment (dongwon8247) [inline gov/governance/errors.gno:25]:
```suggestion
	errBeforeProposalExecutionTime                    = errors.New("[GNOSWAP-GOVERNANCE-016] proposal execution time has not been reached yet")
```

File: launchpad/errors.gno
Comment (dongwon8247) [inline launchpad/errors.gno:17]: Do we block this? I think it's duplicatable via contract

File: staker/errors.gno
Comment (dongwon8247) [inline staker/errors.gno:13]: On what occasion will this occur? Just curious

File: position/errors.gno
Comment (dongwon8247) [inline position/errors.gno:17]: What does this error mean?

File: position/errors.gno
Comment (dongwon8247) [inline position/errors.gno:12]: On what occasion will this occur?

**Review by dongwon8247 — DISMISSED** [review]

File: community_pool/errors.gno
Comment (r3v4s) [reply community_pool/errors.gno:12]: current impl of `grc20reg` doesn't handle errors like above, so for now it needs to be stay.

File: launchpad/errors.gno
Comment (r3v4s) [reply launchpad/errors.gno:17]: to identify project, we use `projectId` which is form of `tokenPath_createHeight`.
therefore it can not be duplicated as view from contract.

File: staker/errors.gno
Comment (r3v4s) [reply staker/errors.gno:13]: current `wugnot` has [minimum amount](https://github.com/gnolang/gno/blob/1a57e81f5e454304dd0de38657d680edef18100d/examples/gno.land/r/demo/wugnot/wugnot.gno#L18-L20) to wrap/unwrap. 004 error is for that.

File: position/errors.gno
Comment (r3v4s) [reply position/errors.gno:17]: same as https://github.com/gnoswap-labs/gnoswap/pull/355#discussion_r1811787519

File: position/errors.gno
Comment (r3v4s) [reply position/errors.gno:12]: same as https://github.com/gnoswap-labs/gnoswap/pull/355#discussion_r1811787519

**Review by dongwon8247 — APPROVED** [review]

**Review by dongwon8247 — APPROVED** [review]

File: router/router.gno
Comment (jinoosss) [inline router/router.gno:251]: This does not appear to be an error in the input.
It appears to be an error out of slippage.

File: router/router.gno
Comment (jinoosss) [inline router/router.gno:260]: This does not appear to be an error in the input.
It appears to be an error out of slippage.

File: router/wrap_unwrap.gno
Comment (jinoosss) [inline router/wrap_unwrap.gno:15]: The function and error handling of unwrap seems to be different.
When unwrap, no error is raised when the quantity is zero.

File: staker/manage_pool_tiers.gno
Comment (jinoosss) [inline staker/manage_pool_tiers.gno:95]: Can you wrap an unknown error in a gnoswap error?

File: staker/manage_pool_tiers.gno
Comment (jinoosss) [inline staker/manage_pool_tiers.gno:180]: Is this not an `errDataNotFound` error?

File: staker/staker.gno
Comment (jinoosss) [inline staker/staker.gno:114]: If there is no liquidity, is it an input error?

**Review by jinoosss — COMMENTED** [review]
Other than a few comments, the messages by condition seem to be well matched.

In my opinion, it would be better if there were rules for error handling at the code level.
For example, the name of a function should be able to predict whether a panic will be thrown or a value will be returned. 
It would also be nice to have rules for layers that throw errors.

**Comment by r3v4s** [comment]
changing to draft.

1. we have ongoing PR that changes core-logic.
- after pr is merged, I'll wrap it up.

File: router/wrap_unwrap.gno
Comment (r3v4s) [reply router/wrap_unwrap.gno:15]: `ugnotAmount` for `func wrap()` is amount of native ugnot send to wrap.
> without senting coin, we can't unwrap it.

`wugnotAmount` for `func unwrap()` is amount of grc20 wugnot to unwrap.
> with 0 amount of wugnot, we can just skip unwrap.

File: staker/manage_pool_tiers.gno
Comment (r3v4s) [reply staker/manage_pool_tiers.gno:180]: function is `setPoolTier` which adds new pool and tier.
Therefore it can not add pre-existing pool.

File: gov/governance/errors.gno
Comment (r3v4s) [reply gov/governance/errors.gno]: d980013b8958f57f6c103a957fd10f4ad52726e2

File: gov/staker/util.gno
Comment (r3v4s) [reply gov/staker/util.gno]: d14806d58ee2897daec3d13f729c5f9cc2c6d2d7

File: pool/pool.gno
Comment (r3v4s) [reply pool/pool.gno]: b4784112e88c0c2027efde3b8b947d3d1fce16a1

File: pool/pool.gno
Comment (r3v4s) [reply pool/pool.gno]: b4784112e88c0c2027efde3b8b947d3d1fce16a1

File: pool/pool.gno
Comment (r3v4s) [reply pool/pool.gno]: b4784112e88c0c2027efde3b8b947d3d1fce16a1

File: pool/pool.gno
Comment (r3v4s) [reply pool/pool.gno]: b4784112e88c0c2027efde3b8b947d3d1fce16a1

File: pool/position.gno
Comment (r3v4s) [reply pool/position.gno]: 68553202e2b4a1eb2cb8fd8264ad6409b5dcf7fe

File: position/position.gno
Comment (r3v4s) [reply position/position.gno]: f2a2b4076c935e1f0adf6e0fa782fee16c9ffd93

File: position/position.gno
Comment (r3v4s) [reply position/position.gno]: f2a2b4076c935e1f0adf6e0fa782fee16c9ffd93

File: router/router.gno
Comment (r3v4s) [reply router/router.gno]: f146719bed956e83b479873fd65361de0fdefaba

File: router/router.gno
Comment (r3v4s) [reply router/router.gno]: f146719bed956e83b479873fd65361de0fdefaba

File: staker/staker.gno
Comment (r3v4s) [reply staker/staker.gno]: 66dfdbc31355c842079738059affb0b6172cd923

**Comment by r3v4s** [comment]
rebased from latest working branch

**Comment by r3v4s** [comment]
ready for review, I've included every changes since when this pr is created.

File: gov/xgns/errors.gno
Comment (notJoon) [inline gov/xgns/errors.gno:14]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

File: launchpad/errors.gno
Comment (notJoon) [inline launchpad/errors.gno:30]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

File: pool/errors.gno
Comment (notJoon) [inline pool/errors.gno:35]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

File: position/errors.gno
Comment (notJoon) [inline position/errors.gno:23]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

File: protocol_fee/errors.gno
Comment (notJoon) [inline protocol_fee/errors.gno:19]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

File: router/errors.gno
Comment (notJoon) [inline router/errors.gno:25]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

File: staker/errors.gno
Comment (notJoon) [inline staker/errors.gno:40]:
```suggestion
	finalErr := ufmt.Errorf("%s || %s", err.Error(), detail)
```

**Review by notJoon — CHANGES_REQUESTED** [review]
Mostly LGTM. I Left some convention things

**Review by dongwon8247 — APPROVED** [review]
please resolve Byeongjun's comments and merge

**Comment by r3v4s** [comment]
> please resolve Byeongjun's comments and merge

Did commit every suggestion made by @notJoon.
Could you approve again?? Sorry for dismissing previous approval.

**Review by dongwon8247 — APPROVED** [review]

**Review by notJoon — APPROVED** [review]

**Pull request merged**
